tests/unit/package_managers/yarn/test_locators.py

```diff
@@ -0,0 +1,574 @@
+    "once-portal@portal:external-packages/once::locator=berryscary%40workspace%3A.",
+    "supports-hyperlinks-folder@file:external-packages/supports-hyperlinks#external-packages/supports-hyperlinks::hash=cfa5f5&locator=berryscary%40workspace%3A.",
+    "strip-ansi-tarball@file:../../external-packages/strip-ansi-4.0.0.tgz::locator=the-answer%40workspace%3Apackages%2Fthe-answer",
+    "strip-ansi-tarball@file:external-packages/strip-ansi-4.0.0.tgz::locator=berryscary%40workspace%3A.",
+    # https dep
+    "c2-wo-deps-2@https://bitbucket.org/cachi-testing/cachi2-without-deps-second/get/09992d418fc44a2895b7a9ff27c4e32d6f74a982.tar.gz",
+    # optional custom patch for a registry dep
+    "left-pad@npm:1.3.0",
+    "left-pad@patch:left-pad@npm%3A1.3.0#~./my-patches/left-pad.patch::version=1.3.0&hash=629bda&locator=berryscary%40workspace%3A.",
+    # optional builtin patch
+    "fsevents@npm:2.3.2",
+    "fsevents@patch:fsevents@npm%3A2.3.2#~builtin<compat/fsevents>::version=2.3.2&hash=df0bf1",
+    # patched patch dependency
+    "fsevents@patch:fsevents@patch%3Afsevents@npm%253A2.3.2%23./my-patches/fsevents.patch%3A%3Aversion=2.3.2&hash=cf0bf0&locator=berryscary%2540workspace%253A.#~builtin<compat/fsevents>::version=2.3.2&hash=df0bf1",
+    # non-optional builtin patch (in reality, the typescript patch is optional)
+    "typescript@npm:5.1.6",
+    "typescript@patch:typescript@npm%3A5.1.6#builtin<compat/typescript>::version=5.1.6&hash=5da071",
+    # multiple patches of all kinds
+    "is-positive@npm:3.1.0",
+    "is-negative@patch:is-positive@npm%3A3.1.0#~builtin<foo>&./my-patches/is-positive.patch&builtin<bar>&~./baz.patch::version=3.1.0&locator=berryscary%40workspace%3A.",
+]
+
+UNSUPPORTED_LOCATORS = [
+    # exec dep
+    "holy-hand-grenade@exec:./generate-holy-hand-grenade.js#./generate-holy-hand-grenade.js::hash=3b5cbd&locator=berryscary%40workspace%3A.",
+    # git deps
+    "c2-wo-deps@https://bitbucket.org/cachi-testing/cachi2-without-deps.git#commit=9e164b97043a2d91bbeb992f6cc68a3d1015086a",
+    "ccto-wo-deps@git@github.com:cachito-testing/cachito-npm-without-deps.git#commit=2f0ce1d7b1f8b35572d919428b965285a69583f6",
+    # specific workspace of a git dep
+    "npm-lifecycle-scripts@https://github.com/chmeliik/js-lifecycle-scripts.git#workspace=my-workspace&commit=0e786c88d5aca79a68428dadaed4b096bf2ae3e0",
+    # patched git dep
+    "ccto-wo-deps@patch:ccto-wo-deps@git@github.com%3Acachito-testing/cachito-npm-without-deps.git%23commit=2f0ce1d7b1f8b35572d919428b965285a69583f6#./.yarn/patches/ccto-wo-deps-git@github.com-e0fce8c89c.patch::version=1.0.0&hash=51a91f&locator=berryscary%40workspace%3A.",
+]
+
+ALL_LOCATORS = SUPPORTED_LOCATORS + UNSUPPORTED_LOCATORS
+
+PARSED_LOCATORS_AND_REFERENCES = [
+    (
+        _ParsedLocator(scope="isaacs", name="cliui", raw_reference="npm:8.0.2"),
+        _ParsedReference(protocol="npm:", source=None, selector="8.0.2", params=None),
```

Comment:
Nitpick: why are we storing the `:` in the protocol? Isn't it simpler to remove it?

Reply:
Probably would, but I figured I would leave the low-level part (_ParsedLocator and _ParsedReference) consistent with yarnberry code